File: Demo/Demo/ViewController.m
//
//  IGRMainViewController.m
//  InstaFilters
//
//  Created by Vitalii Parovishnyk on 12/29/16.
//  Copyright © 2016 IGR Software. All rights reserved.
//

@import IGRFastFilterView;

#import "ViewController.h"

@interface ViewController () <UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@property (nonatomic, weak) IBOutlet IGRFastFilterView *instaFiltersView;

@end

static NSString * const kWorkImageNotification = @"WorkImageNotification";
#define DEMO_IMAGE [UIImage imageNamed:@"demo"]

@implementation ViewController

#pragma mark - Life Cycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self setupTheme];
    
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    [defaultCenter addObserver:self
                      selector:@selector(setupWorkImage:)
                          name:kWorkImageNotification object:nil];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [self setupDemoView];
    });
}

- (void)dealloc
{
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    [defaultCenter removeObserver:self name:kWorkImageNotification object:nil];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
{
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        [self.instaFiltersView layoutSubviews];
    } completion:nil];
    
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
}

#pragma mark - Setter / Getter

- (void)setInstaFiltersView:(IGRFastFilterView *)instaFiltersView
{
    _instaFiltersView = instaFiltersView;
}

#pragma mark - Private

- (void)setupTheme
{
    [IGRFastFilterView appearance].backgroundColor = [UIColor colorWithWhite:0.7 alpha:1.0];
    [IGRFiltersbarCollectionView appearance].backgroundColor = [UIColor colorWithWhite:0.9 alpha:1.0];
}

- (void)setupDemoView
{
    [[NSNotificationCenter defaultCenter] postNotificationName:kWorkImageNotification
                                                        object:DEMO_IMAGE];
}

- (UIImage *)prepareImage
{
    return self.instaFiltersView.processedImage;
}

#pragma mark - Action

- (IBAction)onTouchGetImageButton:(UIBarButtonItem *)sender
{
    UIAlertControllerStyle style = UIAlertControllerStyleActionSheet;
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Select image", @"")
                                                                   message:@""
                                                            preferredStyle:style];
    
    alert.popoverPresentationController.barButtonItem = sender;
    
    __weak typeof(self) weak = self;
    void(^completeActionBlock)(UIImagePickerControllerSourceType) = ^(UIImagePickerControllerSourceType type) {
        UIImagePickerController *pickerView = [[UIImagePickerController alloc] init];
        pickerView.delegate = weak;
        [pickerView setSourceType:type];
        [self presentViewController:pickerView animated:YES completion:nil];
    };
    
    UIAlertAction *action = [UIAlertAction actionWithTitle:NSLocalizedString(@"From Library", @"")
                                                     style:UIAlertActionStyleDefault
                                                   handler:^(UIAlertAction * action) {
                                                       completeActionBlock(UIImagePickerControllerSourceTypePhotoLibrary);
                                                   }];
    [alert addAction:action];
    
    UIAlertAction *cameraAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"From Camera", @"")
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * action) {
                                                             completeActionBlock(UIImagePickerControllerSourceTypeCamera);
                                                         }];
    [alert addAction:cameraAction];
    
    UIAlertAction* cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"")
                                                     style:UIAlertActionStyleCancel
                                                   handler:nil];
				
    [alert addAction:cancel];
    
    [self presentViewController:alert animated:YES completion:nil];
}

- (IBAction)onTouchShareButton:(UIBarButtonItem *)sender
{
    UIImage *image = [self prepareImage];
    
    UIActivityViewController *avc = [[UIActivityViewController alloc] initWithActivityItems:@[image]
                                                                      applicationActivities:nil];
    avc.popoverPresentationController.barButtonItem = sender;
    avc.popoverPresentationController.permittedArrowDirections = UIPopoverArrowDirectionUp;
    
    [self presentViewController:avc animated:YES completion:nil];
}

#pragma mark - NSNotificationCenter

- (void)setupWorkImage:(NSNotification *)notification
{
    NSAssert([notification.object isKindOfClass:[UIImage class]], @"Image only allowed!");
    
    [self.instaFiltersView setImage:notification.object];
}

#pragma mark - PickerDelegates

- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    UIImage *img = [info valueForKey:UIImagePickerControllerOriginalImage];
    [[NSNotificationCenter defaultCenter] postNotificationName:kWorkImageNotification
                                                        object:img];
    
    [picker dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Demo/DemoSwift/ViewController.swift
//
//  ViewController.swift
//  DemoSwift
//
//  Created by Vitalii Parovishnyk on 1/8/17.
//  Copyright © 2017 IGR Software. All rights reserved.
//

import UIKit
import IGRFastFilterView

class ViewController: UIViewController {
    
    @IBOutlet weak fileprivate var instaFiltersView: IGRFastFilterView?
    
    static let kImageNotification = NSNotification.Name(rawValue: "WorkImageNotification")
    let kDemoImage = UIImage.init(named: "demo")
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupTheme()
        
        let notificationCenter = NotificationCenter.default
        notificationCenter.addObserver(self,
                                       selector: #selector(setupWorkImage(_:)),
                                       name: ViewController.kImageNotification,
                                       object: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        DispatchQueue.once(token: "com.igrsoft.fastfilter.demo") {
            self.setupDemoView()
        }
    }
    
    deinit {
        let notificationCenter = NotificationCenter.default
        notificationCenter.removeObserver(self,
                                          name: ViewController.kImageNotification,
                                          object: nil)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        coordinator.animate(alongsideTransition: { (UIViewControllerTransitionCoordinatorContext) in
            self.instaFiltersView?.layoutSubviews()
        }, completion: nil)
    }
    
    fileprivate func setupTheme () {
        IGRFastFilterView.appearance().backgroundColor = UIColor(white:0.7, alpha:1.0)
        IGRFiltersbarCollectionView.appearance().backgroundColor = UIColor(white:0.9, alpha:1.0)
    }
    
    fileprivate func setupDemoView() {
        let notificationCenter = NotificationCenter.default
        notificationCenter.post(name: ViewController.kImageNotification, object: kDemoImage)
    }
    
    func setupWorkImage(_ notification: Notification) {
        assert(notification.object is UIImage, "Image only allowed!");
        instaFiltersView?.setImage(notification.object as! UIImage)
    }
    
    func prepareImage() -> UIImage {
        return self.instaFiltersView!.processedImage!;
    }
    
    @IBAction func onTouchGetImageButton(_ sender: UIBarButtonItem)
    {
        let alert = UIAlertController.init(title: NSLocalizedString("Select image", comment: ""), message: "", preferredStyle: UIAlertControllerStyle.actionSheet)
        
        alert.popoverPresentationController?.barButtonItem = sender;
        
        func completeAlert(type: UIImagePickerControllerSourceType) {
            let pickerView = UIImagePickerController.init()
            pickerView.delegate = self
            pickerView.sourceType = type
            self.present(pickerView, animated: true, completion: nil)
        }
        
        let style = UIAlertActionStyle.default
        var action = UIAlertAction.init(title: NSLocalizedString("From Library", comment: ""),
                                        style: style) { (UIAlertAction) in
                                            completeAlert(type: UIImagePickerControllerSourceType.photoLibrary)
        }
        
        alert.addAction(action)
        
        action = UIAlertAction.init(title: NSLocalizedString("From Camera", comment: ""),
                                    style: style) { (UIAlertAction) in
                                        completeAlert(type: UIImagePickerControllerSourceType.camera)
        }
        
        alert.addAction(action)
        
        action = UIAlertAction.init(title: NSLocalizedString("Cancel", comment: ""),
                                    style: style)
        
        alert.addAction(action)
        
        self.present(alert, animated: true, completion: nil)
    }
    
    
    @IBAction func onTouchShareButton(_ sender: UIBarButtonItem) {
        
        let image = prepareImage()
        
        let avc = UIActivityViewController.init(activityItems: [image], applicationActivities: nil)
        
        avc.popoverPresentationController?.barButtonItem = sender;
        avc.popoverPresentationController?.permittedArrowDirections = .up;
        
        self.present(avc, animated: true, completion: nil)
    }
}

extension ViewController : UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        let image = info[UIImagePickerControllerOriginalImage]
        let notificationCenter = NotificationCenter.default
        notificationCenter.post(name: ViewController.kImageNotification, object: image)
        
        picker.dismiss(animated: true, completion: nil)
    }
}

public extension DispatchQueue {
    
    private static var _onceTracker = [String]()
    
    /**
     Executes a block of code, associated with a unique token, only once.  The code is thread safe and will
     only execute the code once even in the presence of multithreaded calls.
     
     - parameter token: A unique reverse DNS style name such as com.vectorform.<name> or a GUID
     - parameter block: Block to execute once
     */
    public class func once(token: String, block:(Void)->Void) {
        objc_sync_enter(self); defer { objc_sync_exit(self) }
        
        if _onceTracker.contains(token) {
            return
        }
        
        _onceTracker.append(token)
        block()
    }
}

